src/main/java/de/zalando/aruha/nakadi/service/subscription/zk/CuratorZkSubscriptionClient.java

```diff
@@ -0,0 +1,269 @@
+        } catch (final Exception e) {
+            throw new SubscriptionWrappedException(e);
+        }
+    }
+
+    private static byte[] serializeNode(final String session, final String nextSession, final Partition.State state) {
+        return Stream.of(
+                session == null ? "" : session,
+                nextSession == null ? "" : nextSession,
+                state.name()).collect(Collectors.joining(":")).getBytes(CHARSET);
+    }
+
+    private static Partition deserializeNode(final Partition.PartitionKey key, final byte[] data) {
+        final String[] parts = new String(data, CHARSET).split(":");
+        return new Partition(
+                key,
+                parts[0].isEmpty() ? null : parts[0],
+                parts[1].isEmpty() ? null : parts[1],
+                Partition.State.valueOf(parts[2]));
+    }
+
+    @Override
+    public void updatePartitionConfiguration(final Partition partition) {
+        try {
+            log.info("updating partition state: " + partition.getKey() + ":" + partition.getState() + ":" + partition.getSession() + ":" + partition.getNextSession());
+            curatorFramework.setData().forPath(
+                    getPartitionPath(partition.getKey()),
+                    serializeNode(partition.getSession(), partition.getNextSession(), partition.getState()));
+        } catch (final Exception e) {
+            throw new SubscriptionWrappedException(e);
+        }
+    }
+
+    @Override
+    public void incrementTopology() {
+        try {
+            log.info("Incrementing topology version");
+            final Integer curVersion = Integer.parseInt(new String(curatorFramework.getData().forPath(getSubscriptionPath("/topology")), CHARSET));
+            curatorFramework.setData().forPath(getSubscriptionPath("/topology"), String.valueOf(curVersion + 1).getBytes(CHARSET));
+        } catch (final Exception e) {
+            throw new SubscriptionWrappedException(e);
+        }
+    }
+
+    @Override
+    public Partition[] listPartitions() {
+        log.info("fetching partitions information");
+        try {
+            final List<Partition> partitions = new ArrayList<>();
+            for (final String topic : curatorFramework.getChildren().forPath(getSubscriptionPath("/topics"))) {
+                for (final String partition : curatorFramework.getChildren().forPath(getSubscriptionPath("/topics/" + topic))) {
+                    final Partition.PartitionKey key = new Partition.PartitionKey(topic, partition);
+                    partitions.add(deserializeNode(key, curatorFramework.getData().forPath(getPartitionPath(key))));
+                }
+            }
+            return partitions.toArray(new Partition[partitions.size()]);
+        } catch (final Exception e) {
+            throw new SubscriptionWrappedException(e);
+        }
+    }
+
+    @Override
+    public Session[] listSessions() {
```

Comment:
Why have you decided to work with arrays not with lists?
Moreover at the end of this method you actually have list which you have to convert to array.